xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,355 @@
+
+	cancelResource := a.watchResource(rType, resourceName, dWatcher)
+	defer cancelResource()
+
+	// Wait for mgmt server to recv request.
+	<-onStreamRequestDoneCh
+
+	a.resourcesMu.Lock()
+	if a.resources[rType][resourceName].wTimer == nil {
+		t.Fatal("watch resource timer is nil. Want: wTimer to be started.")
+	}
+	if wState := a.resources[rType][resourceName].wState; wState != watchStateRequested {
+		t.Fatalf("watch resource state in: %v. Want: %v", wState, watchStateRequested)
+	}
+	a.resourcesMu.Unlock()
+
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultClientListener(resourceName, "new-rds-resource")},
+		SkipValidation: true,
+	}
+
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Wait for authority to receive the update.
+	<-watcherUpdateCh
+	a.resourcesMu.Lock()
+	if wState := a.resources[rType][resourceName].wState; wState != watchStateReceived {
+		t.Fatalf("watch resource state in: %v. Want: %v", wState, watchStateReceived)
+	}
+	a.resourcesMu.Unlock()
+}
+
+func (s) TestResourceStateTransitionsFromRequestedToStartedOnError(t *testing.T) {
+	// This tests the scenario where the ADS stream breaks after watchResource call.
+	// In this case the timer for the resource also stops and the state is ready
+	// to be restarted later.
+
+	onStreamRequestDoneCh := make(chan struct{})
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	mgmtServerOpts := e2e.ManagementServerOptions{
+		OnStreamRequest: func(i int64, request *v3discoverypb.DiscoveryRequest) error {
+			// This is to avoid calling close() on a closed ch.
+			select {
+			case <-onStreamRequestDoneCh:
+				return nil
+			default:
+			}
+			close(onStreamRequestDoneCh)
+			return nil
+		},
+	}
+	nodeID := uuid.New().String()
+	a, mgmtServer := setupAuthorityWithMgmtServer(ctx, t, mgmtServerOpts, nodeID)
+	defer a.close()
+
+	resourceName := "xdsclient-test-lds-resource"
+	watcherErrCh := make(chan error)
+	watcher := listenerWatcher{resourceName: resourceName, cb: func(update xdsresource.ListenerUpdate, err error) {
+		watcherErrCh <- err
+	}}
+	dWatcher := &delegatingDummyWatcher{watcher: watcher}
+
+	cancelResource := a.watchResource(rType, resourceName, dWatcher)
+	defer cancelResource()
+
+	// Wait for mgmt server to recv request.
+	<-onStreamRequestDoneCh
+
+	a.resourcesMu.Lock()
+	if a.resources[rType][resourceName].wTimer == nil {
+		t.Fatal("watch resource timer is nil. Want: wTimer to be started.")
+	}
+	if wState := a.resources[rType][resourceName].wState; wState != watchStateRequested {
+		t.Fatalf("watch resource state in: %v. Want: %v", wState, watchStateRequested)
+	}
+	a.resourcesMu.Unlock()
+
+	mgmtServer.Stop()
+
+	// Wait for watcher to receive the update.
+	if err := <-watcherErrCh; err == nil {
```

Comment:
Blocking on channel reads in the main test goroutine should be time bound. For example, if the error callback was never invoked in this test case (because of a bug in the code), this channel read would block until the test timeout value passed to the `go test` command which in our case is `7m`. So, instead select on the `ctx.Done()` channel and this channel (the same pattern that we use in most of our other tests).

Reply:
done